### ISSUE TITLE
Moe implementation crashes during fine-tuning

### ISSUE THREAD

**Issue by mzbac** [ml-explore/mlx-examples#323]
Currently, the Moe implementation loops through selected experts per token. This works fine for inference because of the kv cache; it usually only computes one token at a time. However, it causes issues in training because it needs to loop over batch_size*sequence_length. I'm wondering if there is a way to vectorize this computation.

https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/models/mixtral.py#L167-L172

**Comment by awni** [comment]
Definitely there is a way to vectorize it! I [tried this](https://github.com/ml-explore/mlx-examples/blob/e42682dced2d301123da3a260179b5f4bfdfad9e/mixtral/mixtral.py#L132-L141) at one point, but it was very slow. Something like it might work better for training though where batching is more useful

**Comment by awni** [comment]
Though why is it crashing for you? The implicit eval might be an issue 🤔 for training..

**Comment by mzbac** [comment]
> Though why is it crashing for you? The implicit eval might be an issue 🤔 for training..

When I load the quant model and use Lora with the default configuration, the memory usage goes up to over 120GB and eventually gets killed by the operating system. I can see that the model is stuck in a for loop at https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/models/mixtral.py#L168. My assumption is that it will take 4*2048 steps there, and we have 32 layers of the sparse MOE block. The computation required would be 32*4*2048. Perhaps that's why it crashs

**Comment by awni** [comment]
Yes lol that will never work for trianing.. that part of the model was really only designed for inference.. it has to be vectorized to work with training. The implicit eval could still be an issue (it might be slow), but I think it should work.

**Comment by mzbac** [comment]
> Definitely there is a way to vectorize it! I [tried this](https://github.com/ml-explore/mlx-examples/blob/e42682dced2d301123da3a260179b5f4bfdfad9e/mixtral/mixtral.py#L132-L141) at one point, but it was very slow. Something like it might work better for training though where batching is more useful

@awni I tried to replicate the transformers implementation(https://github.com/huggingface/transformers/blob/main/src/transformers/models/mixtral/modeling_mixtral.py#L810-L840), but I got stuck on expert_mask and torch.where. I can't really figure out how to implement it using mx...

**Comment by mzbac** [comment]
> Yes lol that will never work for trianing.. that part of the model was really only designed for inference.. it has to be vectorized to work with training. The implicit eval could still be an issue (it might be slow), but I think it should work.

Okay, I will take a look at the implicit eval and see if I can get it to work.

**Comment by awni** [comment]
@mzbac This implementation loops over expert so it's only 8 linear calls and should be a lot faster: https://github.com/ml-explore/mlx-examples/blob/e42682dced2d301123da3a260179b5f4bfdfad9e/mixtral/mixtral.py#L132-L141

But it still uses an implicit eval.

**Comment by mzbac** [comment]
@awni I have tried using the explicit eval and implemented the code snippet you provided. Now I am able to load the model and calculate the loss. However, I got an error with grad. not sure if it is caused by the explicit eval.

Here is the code that I used for your reference: [lora](https://github.com/mzbac/mlx-examples/blob/mlx_lm/lora/llms/test.py)
Note: The code works for llama and phi2, but not mixtral. I suspect it is caused by the Moe implementation somehow.

**Comment by awni** [comment]
Right, I see you are trying to LoRA fine-tune the gater which requires the gradient through a top-k operation which is not really well defined so we don't currently allow it. (I think for training these models they use the straight-through estimate but then it's important to balance the router.). For LoRA fine-tuning it would be much much simpler to just ignore fine-tuning the gater. Take the expert assignment as given and simply fine-tune the experts.

**Comment by mzbac** [comment]
@awni Thanks for the clarification. I have tried to use lora q,v in the attention layer, but I still get the same error. Do you mean we should only use Lora MLP layer for the mixtral?

**Comment by awni** [comment]
>  I have tried to use lora q,v in the attention layer, but I still get the same error.

Right I think you have to do an explicit [stop gradient](https://ml-explore.github.io/mlx/build/html/python/_autosummary/mlx.core.stop_gradient.html) on the indices otherwise it will try to backprop through it.

**Comment by mzbac** [comment]
@awni I have tried to stop the gradient before the back prop goes to moe layers(https://github.com/mzbac/mlx-examples/blob/mlx_lm/lora/llms/mlx_lm/models/mixtral.py#L260), which makes lora runnable. However, I can't get the loss to decrease. not sure why, maybe I haven't found the correct fine-tune hyperparams yet. I might just start working on moving lora example to mlx-lm and we can put some kind of warning to users if they try to fine-tune mixtral with lora. Please let me know what you think.

**Comment by mzbac** [comment]
> @mzbac I think that's totally fine! We can disallow fine-tuning the MOE style models until we actually get it to work. I can take a look at some point to see if I have any ideas there.
> 
> Regarding moving lora to mlx-lm what do you think about #213 ? I haven't looked in detail yet, but it could be a good starting point.

I will have a look at the PR, and I am thinking not only providing a cli to for lora but also supporting it via Python API (e.g. https://github.com/mzbac/mlx-examples/blob/mlx_lm/lora/llms/test.py) so that the cli arguments doesb't become unmanageable.

**Comment by awni** [comment]
> but also supporting it via Python API

💯 it would be much better to use it in the API as well.

**Comment by mzbac** [comment]
@awni Just wondering if it's possible to prioritize the support for fine-tuning the moe model. Since apple silicon is weak in computation but strong in vram, enabling fine-tuning for the moe model would have huge potential for the community to customize LLM on Apple Silicon specifically for certain domains. It would also contribute back to a wider open-source LLM. Currently, most other platforms require a large amount of vram for fine-tuning moe models, which is not feasible for everyone.

**Comment by awni** [comment]
Yea this would be a good thing to figure out. I can play around with it and see how much effort it would require / what the actual problems are. Without knowing that it's hard to say how difficult the fix will be (and hence properly prioritize).

**Comment by awni** [comment]
@mzbac I tried the Phixtral model (just to try with something faster) and I think it's working. I will make a branch so you can see how I did it.

I ran this command:

```
python lora.py --train --model mlabonne/phixtral-4x2_8 --steps-per-report 5 --lora-layers 8 --batch-size 2 
```

And the validation loss goes down like so:

```
Iter 1: Val loss 3.046, Val took 5.692s
Iter 200: Val loss 1.613, Val took 5.863s
Iter 400: Val loss 1.529, Val took 5.940s
Iter 600: Val loss 1.457, Val took 5.915s
```
Probably with some tuning it could go down a lot faster (# lora layers, rank, scale etc, though I know those are not all yet configurable).

**Comment by awni** [comment]
Here is the branch: https://github.com/ml-explore/mlx-examples/tree/lora_moe

**Comment by mzbac** [comment]
> @mzbac I tried the Phixtral model (just to try with something faster) and I think it's working. I will make a branch so you can see how I did it.
> 
> I ran this command:
> 
> ```
> python lora.py --train --model mlabonne/phixtral-4x2_8 --steps-per-report 5 --lora-layers 8 --batch-size 2 
> ```
> 
> And the validation loss goes down like so:
> 
> ```
> Iter 1: Val loss 3.046, Val took 5.692s
> Iter 200: Val loss 1.613, Val took 5.863s
> Iter 400: Val loss 1.529, Val took 5.940s
> Iter 600: Val loss 1.457, Val took 5.915s
> Iter 800: Val loss 1.418, Val took 5.964s
> ```
> 
> Probably with some tuning it could go down a lot faster (# lora layers, rank, scale etc, though I know those are not all yet configurable).

@awni I haven't tried Phixtral because it wasn't fine-tuned on the gate layer. It always selects expert 0 and 1, so I'm not sure if that's causing the differences. I will take a look at your branch and see if I can apply it to Mixtral. This is currently the main blocker for porting Phixtral to mlx-lm, if it's works I may just start to port Phixtral to mlx-lm.

**Comment by awni** [comment]
Note the location of the [`stop_gradient`](https://github.com/ml-explore/mlx-examples/blob/lora_moe/lora/models/phixtral.py#L125).  I think that's the a noteable difference from what you were doing and what I did. I used the same expert-batched MOE layer. The Phixtral model also has a fused QKV projection, but I don't expect that would account for the difference.

**Comment by awni** [comment]
>  It always selects expert 0 and 1, so I'm not sure if that's causing the differences.

Could you explain this comment? There are 4 experts and it uses a gater to select two of them right?

**Comment by mzbac** [comment]
> > It always selects expert 0 and 1, so I'm not sure if that's causing the differences.
> 
> Could you explain this comment? There are 4 experts and it uses a gater to select two of them right?

Yes, the gate layer will always select the first two experts because it's a merged model using mergekit and hasn't been fine-tuned for the gate layer.

**Comment by awni** [comment]
Interesting! Thanks for the pointer, I did not know that. It seems a bit odd to call it a MOE model if it always uses the same 2 layers ...

**Comment by mzbac** [comment]
> Interesting! Thanks for the pointer, I did not know that. It seems a bit odd to call it a MOE model if it always uses the same 2 layers ...

Yeah, based on my experience with Mixtral lora, it seems that the training loss decreases for a few iterations and then becomes unstable, causing the loss to continuously increase. Maybe the fixed gate layer selection in Phixtral makes it work with lora.

**Comment by awni** [comment]
Yea could be.. I'm trying the 4-bit Nous Hermes Mixtral because I have it on my machine. Which Mixtral model did you try?

**Comment by awni** [comment]
Oddly the loss starts out really high.. seems a bit off 🤔 (I wonder if thats the quantization or the DPO fine-tuning). But it does seem to go down rapidly

```
Iter 1: Val loss 15.067, Val took 17.239s
Iter 5: Train loss 14.295, It/sec 1.433, Tokens/sec 121.479
Iter 10: Train loss 14.222, It/sec 1.137, Tokens/sec 99.584
Iter 15: Train loss 12.953, It/sec 1.058, Tokens/sec 107.488
Iter 20: Train loss 12.522, It/sec 1.099, Tokens/sec 106.196
Iter 25: Train loss 10.472, It/sec 1.103, Tokens/sec 106.521
Iter 30: Train loss 8.861, It/sec 1.015, Tokens/sec 116.285
Iter 35: Train loss 7.487, It/sec 1.009, Tokens/sec 113.371
Iter 40: Train loss 7.373, It/sec 1.097, Tokens/sec 104.645
Iter 45: Train loss 6.439, It/sec 1.102, Tokens/sec 104.454
Iter 50: Train loss 6.402, It/sec 1.075, Tokens/sec 110.267
Iter 55: Train loss 6.029, It/sec 1.094, Tokens/sec 107.630
Iter 60: Train loss 5.787, It/sec 1.144, Tokens/sec 96.332
Iter 65: Train loss 5.618, It/sec 1.072, Tokens/sec 108.090
Iter 70: Train loss 6.138, It/sec 1.034, Tokens/sec 115.629
Iter 75: Train loss 5.873, It/sec 1.085, Tokens/sec 106.810
Iter 80: Train loss 5.529, It/sec 1.062, Tokens/sec 111.693
Iter 85: Train loss 6.325, It/sec 0.941, Tokens/sec 112.678
Iter 90: Train loss 5.494, It/sec 1.079, Tokens/sec 107.899
Iter 95: Train loss 5.356, It/sec 1.100, Tokens/sec 106.516
Iter 100: Train loss 5.453, It/sec 1.097, Tokens/sec 102.505
```

I will try the same Mixtral model you tried if you point me to it.

**Comment by mzbac** [comment]
> Yea could be.. I'm trying the 4-bit Nous Hermes Mixtral because I have it on my machine. Which Mixtral model did you try?

I used the base model and I upload 4 bit quant on the mlx-community space: https://huggingface.co/mlx-community/Mixtral-8x7B-v0.1-hf-4bit-mlx

**Comment by awni** [comment]
Ok I'm running with `mistralai/Mixtral-8x7B-v0.1`, evidently it was in my HF cache 😄

**Comment by awni** [comment]
Unfortunately this one is very slow.. (which is another problem, it should not be so slow). I will let it run for a while and report results in the morning!

**Comment by mzbac** [comment]
> Unfortunately this one is very slow.. (which is another problem, it should not be so slow). I will let it run for a while and report results in the morning!

I will also try your branch from my end on the quant model to see if it fixed the issue.:)

**Comment by awni** [comment]
Here is how it looks so far:

```
Iter 1: Val loss 13.097, Val took 86.755s
Iter 5: Train loss 12.772, It/sec 0.345, Tokens/sec 29.275
Iter 10: Train loss 12.710, It/sec 0.275, Tokens/sec 24.109
Iter 15: Train loss 11.931, It/sec 0.269, Tokens/sec 27.356
Iter 20: Train loss 12.404, It/sec 0.265, Tokens/sec 25.640
Iter 25: Train loss 11.498, It/sec 0.265, Tokens/sec 25.646
Iter 30: Train loss 9.916, It/sec 0.261, Tokens/sec 29.857
Iter 35: Train loss 9.469, It/sec 0.260, Tokens/sec 29.224
Iter 40: Train loss 8.306, It/sec 0.261, Tokens/sec 24.869
Iter 45: Train loss 6.973, It/sec 0.264, Tokens/sec 25.045
Iter 50: Train loss 6.814, It/sec 0.260, Tokens/sec 26.691
Iter 55: Train loss 6.181, It/sec 0.259, Tokens/sec 25.441
Iter 60: Train loss 5.878, It/sec 0.263, Tokens/sec 22.153
Iter 65: Train loss 5.871, It/sec 0.263, Tokens/sec 26.478
Iter 70: Train loss 6.278, It/sec 0.260, Tokens/sec 29.022
Iter 75: Train loss 6.094, It/sec 0.259, Tokens/sec 25.512
Iter 80: Train loss 5.713, It/sec 0.260, Tokens/sec 27.319
Iter 85: Train loss 6.168, It/sec 0.259, Tokens/sec 31.071
Iter 90: Train loss 5.831, It/sec 0.260, Tokens/sec 26.004
Iter 95: Train loss 5.765, It/sec 0.255, Tokens/sec 24.669
Iter 100: Train loss 5.823, It/sec 0.261, Tokens/sec 24.334
Iter 100: Saved adapter weights to adapters.npz.
Iter 105: Train loss 5.594, It/sec 0.258, Tokens/sec 27.985
Iter 110: Train loss 5.769, It/sec 0.258, Tokens/sec 27.899
Iter 115: Train loss 5.208, It/sec 0.254, Tokens/sec 26.616
Iter 120: Train loss 5.451, It/sec 0.259, Tokens/sec 20.025
Iter 125: Train loss 5.415, It/sec 0.258, Tokens/sec 25.544
Iter 130: Train loss 5.359, It/sec 0.261, Tokens/sec 23.209
Iter 135: Train loss 5.935, It/sec 0.256, Tokens/sec 26.165
Iter 140: Train loss 5.326, It/sec 0.256, Tokens/sec 27.596
Iter 145: Train loss 5.275, It/sec 0.254, Tokens/sec 23.596
```

- The loss starts out very high which is suspicious, I should try generating text with it to make sure it is working, maybe there is a bug in the batching code..
- But it does go down which is good..
- Time is way too slow. The quantized model is 4x faster which makes no sense for training...and it's not swapping so I don't know what's going on there.

**Comment by mzbac** [comment]
> Here is how it looks so far:
> 
> ```
> Iter 1: Val loss 13.097, Val took 86.755s
> Iter 5: Train loss 12.772, It/sec 0.345, Tokens/sec 29.275
> Iter 10: Train loss 12.710, It/sec 0.275, Tokens/sec 24.109
> Iter 15: Train loss 11.931, It/sec 0.269, Tokens/sec 27.356
> Iter 20: Train loss 12.404, It/sec 0.265, Tokens/sec 25.640
> Iter 25: Train loss 11.498, It/sec 0.265, Tokens/sec 25.646
> Iter 30: Train loss 9.916, It/sec 0.261, Tokens/sec 29.857
> Iter 35: Train loss 9.469, It/sec 0.260, Tokens/sec 29.224
> Iter 40: Train loss 8.306, It/sec 0.261, Tokens/sec 24.869
> Iter 45: Train loss 6.973, It/sec 0.264, Tokens/sec 25.045
> Iter 50: Train loss 6.814, It/sec 0.260, Tokens/sec 26.691
> Iter 55: Train loss 6.181, It/sec 0.259, Tokens/sec 25.441
> Iter 60: Train loss 5.878, It/sec 0.263, Tokens/sec 22.153
> Iter 65: Train loss 5.871, It/sec 0.263, Tokens/sec 26.478
> Iter 70: Train loss 6.278, It/sec 0.260, Tokens/sec 29.022
> Iter 75: Train loss 6.094, It/sec 0.259, Tokens/sec 25.512
> Iter 80: Train loss 5.713, It/sec 0.260, Tokens/sec 27.319
> Iter 85: Train loss 6.168, It/sec 0.259, Tokens/sec 31.071
> Iter 90: Train loss 5.831, It/sec 0.260, Tokens/sec 26.004
> Iter 95: Train loss 5.765, It/sec 0.255, Tokens/sec 24.669
> Iter 100: Train loss 5.823, It/sec 0.261, Tokens/sec 24.334
> Iter 100: Saved adapter weights to adapters.npz.
> Iter 105: Train loss 5.594, It/sec 0.258, Tokens/sec 27.985
> Iter 110: Train loss 5.769, It/sec 0.258, Tokens/sec 27.899
> Iter 115: Train loss 5.208, It/sec 0.254, Tokens/sec 26.616
> Iter 120: Train loss 5.451, It/sec 0.259, Tokens/sec 20.025
> Iter 125: Train loss 5.415, It/sec 0.258, Tokens/sec 25.544
> Iter 130: Train loss 5.359, It/sec 0.261, Tokens/sec 23.209
> Iter 135: Train loss 5.935, It/sec 0.256, Tokens/sec 26.165
> Iter 140: Train loss 5.326, It/sec 0.256, Tokens/sec 27.596
> Iter 145: Train loss 5.275, It/sec 0.254, Tokens/sec 23.596
> ```
> 
> * The loss starts out very high which is suspicious, I should try generating text with it to make sure it is working, maybe there is a bug in the batching code..
> * But it does go down which is good..
> * Time is way too slow. The quantized model is 4x faster which makes no sense for training...and it's not swapping so I don't know what's going on there.

@awni That may be caused by the current mixtral imp only returning the last token. We need to remove the slicing -> https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/models/mixtral.py#L232

**Comment by mzbac** [comment]
@awni your implementation works. I have tested it on mixtral qlora (with gate) and it works well. create this [PR](https://github.com/ml-explore/mlx-examples/pull/343) to add it to the lora example.

**Comment by awni** [comment]
Good catch regarding the last token 🤦‍♂️
